### PR TITLE
New expression for Physics: LinearVelocityAngle()

### PR DESCRIPTION
**Pull request by tristanbob**
This expression returns the angle that a physics object is moving. It is similar to the ForceAngle() expression.

## Questions: 
- What should this function be named? Currently: `LinearVelocityAngle()`
- The output range is currently: (-180,180).  Should I add `180` to this expression so users see a range of (0,360)?

## Playable game
https://liluo.io/games/b12d2d8d-18eb-4a79-8e71-c60565dad547?dev=true

## Project files
[LinearVelocityAngle.zip](https://github.com/4ian/GDevelop/files/10365579/LinearVelocityAngle.zip)

## Video

https://user-images.githubusercontent.com/8879811/211133282-b58b1e09-ee3c-4405-8c21-9700078dfaa0.mp4

### CODE REVIEW

**Comment by arthuro555** [comment]
> What should this function be named? Currently: LinearVelocityAngle()

That'd be OK imo, but a name like `MovementDirection` might be more beginner friendly, albeit less precise

> The output range is currently: (-180,180). Should I add 180 to this expression so users see a range of (0,360)?

I think it's fine as it is. If it was important to do that, `gdjs.toDegrees` would likely do that for you already. 

---

Your code fails automated tests because it doesn't respect the formatting rules: run `npm run format` in `/GDJS` to automatically format the code according to our formatting rules, commit & push ;)

**Comment by tristanbob** [comment]
> That'd be OK imo, but a name like `MovementDirection` might be more beginner friendly, albeit less precise

I don't like my name much. I'd love to follow a similar format to other functions, but we have a wide range:

### Non-physics forces:
- ForceX()
- ForceY()
- ForceLength()
- ForceAngle()


### Physics:
- LinearVelocityX()
- LinearVelocityY()
- LinearVelocityLength()
- ???

> > The output range is currently: (-180,180). Should I add 180 to this expression so users see a range of (0,360)?
> 
> I think it's fine as it is. If it was important to do that, `gdjs.toDegrees` would likely do that for you already.

I'm not sure what range is best.
BTW, I am using ToDegress to convert from Radians. It doesn't adjust the range of degrees.

> Your code fails automated tests because it doesn't respect the formatting rules: run `npm run format` in `/GDJS` to automatically format the code according to our formatting rules, commit & push ;)

Thanks, I will do this now!  (I have a lot to learn about core contributions)

**Comment by 4ian** [comment]
Thank you @tristanbob! 
Can I be a pain and ask you to revert to the original naming?

Physics:
- LinearVelocityX()
- LinearVelocityY()
- LinearVelocityLength()
- LinearVelocityAngle()

as this will ensure consistency inside the behavior and with the builtin forces.

"MovementAngle" is also something a bit "dangerous" because very generic. It's the angle of the movement... but what if the object just got also slightly teleported by something else outside the physics engine. Is this then the proper movement angle? What movement is this?

I agree that Linear Velocity is not super beginner friendly, but at least it's 100% correct. If we find something better, we could always add a synonym later. But at least we'll keep the generic "movement angle" name free for another abstraction or something else later :)

**Comment by D8H** [comment]
Should Physics 2 behavior follows the pathfinding and top-down movement behaviors and exposes an angle condition with a tolerance parameter to avoid users to care about the actual angle value (-90° = 270°)?

**Comment by 4ian** [comment]
> Should Physics 2 behavior follows the pathfinding and top-down movement behaviors and exposes an angle condition with a tolerance parameter to avoid users to care about the actual angle value (-90° = 270°)?

Probably a good idea. But this can be another condition in addition to this one?

**Comment by D8H** [comment]
> > Should Physics 2 behavior follows the pathfinding and top-down movement behaviors and exposes an angle condition with a tolerance parameter to avoid users to care about the actual angle value (-90° = 270°)?
> 
> Probably a good idea. But this can be another condition in addition to this one?

I think the old one should be hidden because it misleads users into comparing 2 angles like this `angle1 = angle2` which make little sense. The only angle that can be useful to compare as a number is the object angle because a meaning can be given to angles making more than one cycle (but this should still be a very rare case).

I think a "compare 2 angles with tolerance" condition should be added next to the "compare 2 numbers" to guide users. They may not know how to use `AngleDifference()` or even it exists.

To sum up, movement angles should never be compare as numbers.

**Comment by 4ian** [comment]
> I think a "compare 2 angles with tolerance" condition should be added next to the "compare 2 numbers" to guide users. > They may not know how to use AngleDifference() or even it exists.

Yes, totally.

> To sum up, movement angles should never be compared as numbers.

I'm not entirely sure, you might want to do something when the movement angle has gone above a certain specific number?

**Comment by D8H** [comment]
> > To sum up, movement angles should never be compared as numbers.
> 
> I'm not entirely sure, you might want to do something when the movement angle has gone above a certain specific number?

I guess if the player must never go left, you can check -135° with 45° for one side and +123° with 45° for the other side. I don't think there is a notion of "above" for a movement angle: is 180° above -180°?

**Review by 4ian — APPROVED** [review]

**Comment by tristanbob** [comment]
FYI - I tested this expression in the new release and it worked great.
